core/dbt/adapters/base/connections.py

```diff
@@ -159,6 +176,88 @@ def set_connection_name(self, name: Optional[str] = None) -> Connection:
         conn.name = conn_name
         return conn
 
+    @classmethod
+    def retry_connection(
+        cls,
+        connection: Connection,
+        connect: Callable[[], AdapterHandle],
+        logger: AdapterLogger,
+        retryable_exceptions: Iterable[Type[Exception]],
+        retry_limit: int = 1,
+        retry_timeout: Union[Callable[[int], SleepTime], SleepTime] = 1,
```

Comment:
The type signature is definitely simpler, and, although exp. backoff could work with some state that the caller has (as it knows the max. retries when calling `retry_connection`), it's not very intuitive. An increasing `int` feels right.

Personally, I lean towards having a `attempts: int = 0` + docs just because it's simpler to implement. Perhaps with a name to indicate that it's an internal counter: `attempt_counter: int = 0`, `internal_attempts: int = 0`, or an underscore `_attempts: int = 0`.

Reply:
That being said, I'm not sure I like the removal of the option to just pass an int/float. Sure the type signature is simpler, but from a caller perspective, `lambda: 1` is not intuitive. Perhaps we could go with different arguments?

For example:

```
    @classmethod
    def retry_connection(
        cls,
        connection: Connection,
        connect: Callable[[], AdapterHandle],
        logger: AdapterLogger,
        retryable_exceptions: Iterable[Type[Exception]],
        retry_limit: int = 1,
        retry_timeout: SleepTime = 1,
        retry_timeout_func: Optional[Callable[[int], SleepTime]] = None,
    ) -> Connection:
        timeout = retry_timeout_func(...) if retry_timeout_func else retry_timeout
```

Reply:
for the sake of time, let's just add in `_attempts: int = 0`. Keeping the type a union of callable and sleeptime is fine.

Reply:
It was added in latest commit.